File: src/main/domain/exceptions.ts
/** Domain exceptions for Azure Repos Scan. */

export class DomainError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "DomainError";
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class AuthenticationError extends DomainError {
  constructor(
    message: string = "Autenticación fallida con Azure DevOps",
  ) {
    super(message);
    this.name = "AuthenticationError";
  }
}

export class OrganizationNotFoundError extends DomainError {
  constructor(org: string) {
    super(`Organización no encontrada: ${org}`);
    this.name = "OrganizationNotFoundError";
  }
}

export class ApiError extends DomainError {
  readonly statusCode: number;
  readonly detail: string;

  constructor(statusCode: number, detail: string) {
    super(`Error API (${statusCode}): ${detail}`);
    this.name = "ApiError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
}

export class RateLimitError extends DomainError {
  readonly retryAfter: number;

  constructor(retryAfter: number) {
    super(`Rate limit alcanzado. Reintentar en ${retryAfter}s`);
    this.name = "RateLimitError";
    this.retryAfter = retryAfter;
  }
}

export class SearchIndexNotReadyError extends DomainError {
  constructor() {
    super(
      "El índice de Code Search no está listo o no está habilitado",
    );
    this.name = "SearchIndexNotReadyError";
  }
}

File: src/azure_repos_scan/domain/exceptions.py
"""Excepciones del dominio."""

from __future__ import annotations

from typing import final


class DomainError(Exception):
    """Excepción base del dominio."""


@final
class AuthenticationError(DomainError):
    """Error de autenticación con Azure DevOps."""

    def __init__(self, message: str = "Autenticación fallida con Azure DevOps") -> None:
        super().__init__(message)


@final
class OrganizationNotFoundError(DomainError):
    """Organización no encontrada."""

    def __init__(self, org: str) -> None:
        super().__init__(f"Organización no encontrada: {org}")


@final
class ApiError(DomainError):
    """Error genérico de la API de Azure DevOps."""

    def __init__(self, status_code: int, detail: str) -> None:
        self.status_code = status_code
        self.detail = detail
        super().__init__(f"Error API ({status_code}): {detail}")


@final
class RateLimitError(DomainError):
    """Límite de tasa alcanzado en la API."""

    def __init__(self, retry_after: int) -> None:
        self.retry_after = retry_after
        super().__init__(f"Rate limit alcanzado. Reintentar en {retry_after}s")


@final
class SearchIndexNotReadyError(DomainError):
    """El índice de búsqueda no está listo."""

    def __init__(self) -> None:
        super().__init__("El índice de Code Search no está listo o no está habilitado")
